text/3703-wrapping-from.md

```diff
@@ -0,0 +1,277 @@
+
+[ACP#453]: https://github.com/rust-lang/libs-team/issues/453
+
+## What about using `From<Wrapping<T>>` instead of a new trait?
+
+Coherent makes this one less appetizing.
+
+With a new trait, one can do this outside `core`
+```rust
+impl WrappingFrom<&BigInteger> for u32 {
+    fn wrapping_from(x: &BigInteger) -> u32 { … }
+}
+```
+
+whereas with `num::Wrapping` you get an error like
+```text
+error[E0117]: only traits defined in the current crate can be implemented for primitive types
+ --> src/lib.rs:4:1
+  |
+4 | impl From<Wrapping<&BigInteger>> for u32 {
+  | ^^^^^---------------------------^^^^^---
+  | |    |                               |
+  | |    |                               `u32` is not defined in the current crate
+  | |    `Wrapping` is not defined in the current crate
+  | impl doesn't use only types from inside the current crate
+  |
+  = note: define and implement a trait or new type instead
+```
+
+Maybe it'd be possible to mark `Wrapping` as `#[fundamental]` and thus allow crates to add impls like that,
+but that's a much bigger hammer than just adding a new trait for it.
+
+Also, having a separate trait that's just about numerics means that the error message when something isn't implemented
+can talk just about the implementations for that trait, rather than potentially giving you the giant list of every
+`From` that probably includes a bunch of irrelevant ones.  (That said, smart diagnostics could mitigate this too.)
+
+## Can we implement this for `NonZero<_>`?
+
+Trivially there cannot be a blanket impl
+```rust
+impl<T: WrappingFrom<U>, U> WrappingFrom<NonZero<U>> for NonZero<T> { … }
+```
+because `NonZero<u8>::wrapping_from(NonZero(0x100_u16))` would give `0`, which is disallowed.
+
+Thus what, if anything, to do here is a philosophical question about the intent of `NonZero` and `WrappingFrom`.
+
+This RFC doesn't propose any implementations for `NonZero` as part of the initial change.  People can use
+```rust
+NonZero::new(u8::wrapping_from(x.get()))
+```
+or similar if they want the version that gives an `Option`, which seems fine for now.
+
+There's at least two possible way we could go here:
+
+1. Convert to `Option`
+
+We could plausibly have something like
+```rust
+impl<T: WrappingFrom<U>, U> WrappingFrom<NonZero<U>> for Option<NonZero<T>> { … }
+```
+which calls `NonZero::new` internally.
+
+That seems like it'd be annoying to use, though, since the caller would have
+to write out the `Option::<NonZero<u8>>::wrapping_from(blah)`.
+
+2. Use the non-zero lattice instead
+
+It turns out that
+- `NonZero<u8>` has 2⁸ - 1 = 255 × 1 values
+- `NonZero<u16>` has 2¹⁶ - 1 = 255 × 257 values
+- `NonZero<u32>` has 2³² - 1 = 255 × 16843009 values
+- `NonZero<u64>` has 2⁶⁴ - 1 = 255 × 72340172838076673 values
+- `NonZero<u128>` has 2¹²⁸ - 1 = 255 × 1334440654591915542993625911497130241 values
+
+(This works for anything that's a multiple of octets, as `0xFF…FF = 0xFF * 0x01…01`.)
```

Comment:
It works for more than just factors of `0xFF`.
also, you should use latex math syntax, since that does work on github and doesn't use a bunch of unicode characters that may not render properly (e.g. they didn't for me with the monospace font on my phone). Rendered example:
- `NonZero<u128>` has $2^{128} - 1 = \mathrm{0xFF} \times \mathrm{0x101} \times \mathrm{0x10001} \times \mathrm{0x100000001} \times \mathrm{0x10000000000000001}$ values
```suggestion
It turns out that
- `NonZero<u8>` has $2^8 - 1 = \mathrm{0xFF}$ values
- `NonZero<u16>` has $2^{16} - 1 = \mathrm{0xFF} \times \mathrm{0x101}$ values
- `NonZero<u32>` has $2^{32} - 1 = \mathrm{0xFF} \times \mathrm{0x101} \times \mathrm{0x10001}$ values
- `NonZero<u64>` has $2^{64} - 1 = \mathrm{0xFF} \times \mathrm{0x101} \times \mathrm{0x10001} \times \mathrm{0x100000001}$ values
- `NonZero<u128>` has $2^{128} - 1 = \mathrm{0xFF} \times \mathrm{0x101} \times \mathrm{0x10001} \times \mathrm{0x100000001} \times \mathrm{0x10000000000000001}$ values

(This works for anything where the number of bits is a multiple of the number of bits of a smaller type, as `0xFF…FF = 0xFF * 0x01…01` and so on for larger groups of `F`s.)
```